In [1]:
import pandas as pd
customers = pd.read_csv("../01_raw_data/olist_customers_dataset.csv")
customers.head()
customers.head(10)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP
5,879864dab9bc3047522c92c82e1212b8,4c93744516667ad3b8f1fb645a3116a4,89254,jaragua do sul,SC
6,fd826e7cf63160e536e0908c76c3f441,addec96d2e059c80c30fe6871d30d177,4534,sao paulo,SP
7,5e274e7a0c3809e14aba7ad5aae0d407,57b2a98a409812fe9618067b6b8ebe4f,35182,timoteo,MG
8,5adf08e34b2e993982a47070956c5c65,1175e95fb47ddff9de6b2b06188f7e0d,81560,curitiba,PR
9,4b7139f34592b3a31687243a302fa75b,9afe194fb833f79e300e37e580171f22,30575,belo horizonte,MG


In [2]:
customers.shape

(99441, 5)

In [3]:
customers.columns

Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='object')

In [4]:
#data types
customers.dtypes

customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

In [5]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [6]:
customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [7]:
customers.duplicated().sum()

0

In [8]:
customers["customer_id"].duplicated().sum()

0

In [9]:
#extra repeated records
customers["customer_unique_id"].duplicated().sum()

3345

In [10]:
customers["customer_unique_id"].nunique()

96096

In [11]:
customers["customer_unique_id"].value_counts().head(10)

customer_unique_id
8d50f5eadf50201ccdcedfb9e2ac8455    17
3e43e6105506432c953e165fb2acf44c     9
1b6c7548a2a1f9037c1fd3ddfed95f33     7
ca77025e7201e3b30c44b472ff346268     7
6469f99c1f9dfae7733b25662e7f1782     7
63cfc61cee11cbe306bff5857d00bfe4     6
47c1a3033b8b77b3ab6e109eb4d5fdf3     6
12f5d6e1cbf93dafd9dcc19095df0b3d     6
de34b16117594161a6a89c50b289d35a     6
dc813062e0fc23409cd255f7f53c7074     6
Name: count, dtype: int64

In [12]:
# calculating the number of actual customers who purchased more than once repeat customer
(customers["customer_unique_id"].value_counts() > 1).sum()

2997

In [13]:
repeat_customers = (customers["customer_unique_id"].value_counts() > 1) .sum()
unique_customers = customers["customer_unique_id"].nunique()
repeat_customer_rate = (repeat_customers / unique_customers) * 100
print(round(repeat_customer_rate,2))

3.12


In [14]:
customers["customer_state"].value_counts()

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: count, dtype: int64

In [15]:
customers["customer_state"].nunique()

27

In [16]:
customers["customer_city"].nunique()

4119

In [17]:
customers["customer_city"].value_counts().head(10)

customer_city
sao paulo                15540
rio de janeiro            6882
belo horizonte            2773
brasilia                  2131
curitiba                  1521
campinas                  1444
porto alegre              1379
salvador                  1245
guarulhos                 1189
sao bernardo do campo      938
Name: count, dtype: int64

In [18]:
customers["customer_city"].sort_values().head(20)

44027    abadia dos dourados
74529    abadia dos dourados
19445    abadia dos dourados
42307              abadiania
5205                  abaete
11734                 abaete
49712                 abaete
61140                 abaete
61789                 abaete
41957                 abaete
77482                 abaete
52849                 abaete
82676                 abaete
24154                 abaete
47918                 abaete
53051                 abaete
44867             abaetetuba
11200             abaetetuba
68477             abaetetuba
93641             abaetetuba
Name: customer_city, dtype: object

In [19]:
# checking whether any city has space at the beginning or the ending
(customers["customer_city"] != customers["customer_city"].str.strip()).sum() 

0

In [20]:
# checking whether any state 2 character long
customers["customer_state"].str.len().value_counts()

customer_state
2    99441
Name: count, dtype: int64

In [21]:
customers["customer_zip_code_prefix"].astype(str).str.len().value_counts()

customer_zip_code_prefix
5    75446
4    23995
Name: count, dtype: int64

this means 75,446 rows currently have a postcode prefix that looks 5 digits long
23,995 rows currently have a postcode prefix that looks 4 digits long

In [22]:
orders = pd.read_csv("../01_raw_data/olist_orders_dataset.csv")
orders.shape

(99441, 8)

In [23]:
orders.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [24]:
orders.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')

In [25]:
orders.dtypes

order_id                         object
customer_id                      object
order_status                     object
order_purchase_timestamp         object
order_approved_at                object
order_delivered_carrier_date     object
order_delivered_customer_date    object
order_estimated_delivery_date    object
dtype: object

In [26]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [27]:
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [28]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [29]:
orders[
    (orders["order_status"] == "delivered")&
    (orders["order_delivered_customer_date"].isnull())
]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaN,2017-12-18 00:00:00
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaN,2018-07-16 00:00:00
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaN,2018-07-30 00:00:00
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaN,2018-07-24 00:00:00
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaN,NaN,2017-06-23 00:00:00
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaN,2018-06-26 00:00:00
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaN,2018-07-19 00:00:00


In [30]:
#This gives the number of rows matching both conditions:
#order_status == "delivered"
#delivery date is missing
orders[
    (orders["order_status"] == "delivered")&
    (orders["order_delivered_customer_date"].isnull())
].shape[0]

8

8 orders are marked as delivered but have no recorded customer delivery date.

In [31]:
orders[
    (orders["order_status"] != "delivered")&
    (orders["order_delivered_customer_date"].notnull())
]["order_status"].value_counts()

order_status
canceled    6
Name: count, dtype: int64

6 orders marked as a cancelled even though they have delivery date

In [32]:
orders["order_purchase_timestamp"] = pd.to_datetime(
    orders["order_purchase_timestamp"]
)
orders["order_approved_at"] = pd.to_datetime(
    orders["order_approved_at"]
)
(
    orders["order_approved_at"]<
    orders["order_purchase_timestamp"]
).sum()

0

There are no orders approved before they were purchased

In [33]:
orders["order_delivered_carrier_date"] = pd.to_datetime(
    orders["order_delivered_carrier_date"]
)
orders["order_delivered_customer_date"] = pd.to_datetime(
    orders["order_delivered_customer_date"]
)
orders["order_estimated_delivery_date"] = pd.to_datetime(
    orders["order_estimated_delivery_date"]
)

In [34]:
(
    orders["order_delivered_customer_date"]<
    orders["order_delivered_carrier_date"]
).sum()

23

The customer received the parcel before the parcel was handed to the customer is 23

In [35]:
orders[
    orders["order_delivered_customer_date"]<
    orders["order_delivered_carrier_date"]
][[
    "order_id",
    "order_status",
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date"
]].head(10)

,order_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date
6437,a1abeb653a4d4cd1e142ccb8c82cd069,delivered,2017-07-20 11:20:52,2017-07-21 06:43:14,2017-07-28 16:57:58,2017-07-25 19:32:56
9553,383aa8b2724fe452d9ccd9934a8c628b,delivered,2017-07-02 20:58:43,2017-07-02 21:10:20,2017-07-07 17:22:41,2017-07-06 14:27:51
13487,cb1134f9010d242e9515ad1c78ec0c39,delivered,2017-07-16 12:35:34,2017-07-18 06:03:50,2017-07-20 19:22:02,2017-07-19 14:13:28
14474,dceb62e8fa94b46006c9554fed743df0,delivered,2017-07-20 20:58:05,2017-07-22 11:45:11,2017-08-01 18:23:30,2017-07-26 18:09:10
19268,5f9d46795c3126674e52becb3a1a517f,delivered,2017-07-18 11:48:20,2017-07-18 12:03:29,2017-07-20 23:03:42,2017-07-20 18:52:41
21338,8c78d01de3a9009e23d6877a7cc9be20,delivered,2016-10-08 15:36:50,2016-10-08 18:13:44,2016-10-26 11:41:53,2016-10-25 17:51:46
22520,b27af682321527a6349f1761eb3f360c,delivered,2017-06-14 20:17:04,2017-06-14 20:30:08,2017-06-27 14:51:54,2017-06-26 15:45:35
25393,1cc3ae63caffff2d6c3ee3e78e074acf,delivered,2017-08-07 21:35:22,2017-08-08 21:45:15,2017-08-10 18:28:56,2017-08-10 18:05:38
25646,e37f11cae9985ca58f0b56f268720537,delivered,2017-07-26 11:46:34,2017-07-27 10:10:16,2017-08-01 18:17:47,2017-07-31 17:49:56
27470,fa3e37584f4fdb1ded0e0de700dfcb4e,delivered,2017-07-30 19:32:23,2017-07-30 19:45:09,2017-08-09 18:18:43,2017-08-01 21:13:01


note : 23 orders have inconsistent delivery timelines where customer delivery is recorded before career handover

In [36]:
suspicious_orders = orders[
    orders["order_delivered_customer_date"]<
    orders["order_delivered_carrier_date"]
].copy()

suspicious_orders["delivery_date_gap"] = (
    suspicious_orders["order_delivered_carrier_date"] - 
    suspicious_orders["order_delivered_customer_date"]
)
suspicious_orders[[
    "order_id",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "delivery_date_gap"
]].head(10)

,order_id,order_delivered_carrier_date,order_delivered_customer_date,delivery_date_gap
6437,a1abeb653a4d4cd1e142ccb8c82cd069,2017-07-28 16:57:58,2017-07-25 19:32:56,2 days 21:25:02
9553,383aa8b2724fe452d9ccd9934a8c628b,2017-07-07 17:22:41,2017-07-06 14:27:51,1 days 02:54:50
13487,cb1134f9010d242e9515ad1c78ec0c39,2017-07-20 19:22:02,2017-07-19 14:13:28,1 days 05:08:34
14474,dceb62e8fa94b46006c9554fed743df0,2017-08-01 18:23:30,2017-07-26 18:09:10,6 days 00:14:20
19268,5f9d46795c3126674e52becb3a1a517f,2017-07-20 23:03:42,2017-07-20 18:52:41,0 days 04:11:01
21338,8c78d01de3a9009e23d6877a7cc9be20,2016-10-26 11:41:53,2016-10-25 17:51:46,0 days 17:50:07
22520,b27af682321527a6349f1761eb3f360c,2017-06-27 14:51:54,2017-06-26 15:45:35,0 days 23:06:19
25393,1cc3ae63caffff2d6c3ee3e78e074acf,2017-08-10 18:28:56,2017-08-10 18:05:38,0 days 00:23:18
25646,e37f11cae9985ca58f0b56f268720537,2017-08-01 18:17:47,2017-07-31 17:49:56,1 days 00:27:51
27470,fa3e37584f4fdb1ded0e0de700dfcb4e,2017-08-09 18:18:43,2017-08-01 21:13:01,7 days 21:05:42


In [37]:
suspicious_orders["delivery_date_gap"].max()

Timedelta('16 days 02:18:29')

In [38]:
suspicious_orders["delivery_date_gap"].mean()

Timedelta('3 days 06:27:18.478260869')

In [39]:
late_orders[[
    "order_delivered_customer_date",
    "order_estimated_delivery_date",
    "days_late"
]].head(10)

NameError: name 'late_orders' is not defined

In [40]:
late_orders = orders[
    (orders["order_status"] == "delivered")&
    (orders["order_delivered_customer_date"].notnull())&
    (
        orders["order_delivered_customer_date"].dt.normalize() >
        orders["order_estimated_delivery_date"].dt.normalize()
    )
].copy()

late_orders.shape

(6534, 8)

In [41]:
late_delivery_rate = (
    late_orders.shape[0]/delivered_orders
)*100
round(late_delivery_rate,2)

NameError: name 'delivered_orders' is not defined

In [42]:
late_orders["days_late"] = (
    late_orders["order_delivered_customer_date"].dt.normalize() - 
    late_orders["order_estimated_delivery_date"].dt.normalize()
).dt.days
late_orders["days_late"].describe()

count    6534.000000
mean       10.620141
std        14.644955
min         1.000000
25%         3.000000
50%         7.000000
75%        13.000000
max       188.000000
Name: days_late, dtype: float64

In [43]:
late_orders[
    [
        "order_id",
        "order_purchase_timestamp",
        "order_estimated_delivery_date",
        "order_delivered_customer_date",
        "days_late"
    ]
].sort_values(
    by = "days_late",
    ascending = False
).head(10)

,order_id,order_purchase_timestamp,order_estimated_delivery_date,order_delivered_customer_date,days_late
55619,1b3190b2dfa9d789e1f14c05b647a14a,2018-02-23 14:57:35,2018-03-15,2018-09-19 23:24:07,188
19590,ca07593549f1816d26a572e06dc1eab6,2017-02-21 23:31:27,2017-03-22,2017-09-19 14:36:39,181
11399,47b40429ed8cce3aee9199792275433f,2018-01-03 09:44:01,2018-01-19,2018-07-13 20:51:31,175
81401,2fe324febf907e3ea3f2aa9650869fa5,2017-03-13 20:17:10,2017-04-05,2017-09-19 17:00:07,167
89130,285ab9426d6982034523a855f55a885e,2017-03-08 22:47:40,2017-04-06,2017-09-19 14:00:04,166
61610,440d0d17af552815d15a9e41abe49359,2017-03-07 23:59:51,2017-04-07,2017-09-19 15:12:50,165
68769,c27815f7e3dd0b926b58552628481575,2017-03-15 23:23:17,2017-04-10,2017-09-19 17:14:25,162
40847,d24e8541128cea179a11a65176e0a96f,2017-06-12 13:14:11,2017-06-26,2017-12-04 18:36:29,161
38509,0f4519c5f1c541ddec9f21b3bddd533a,2017-03-09 13:26:57,2017-04-11,2017-09-19 14:38:21,161
54480,2d7561026d542c8dbd8f0daeadf67a43,2017-03-15 11:24:27,2017-04-13,2017-09-19 14:38:18,159


In [44]:
(
    late_orders["order_delivered_customer_date"].dt.date ==
    pd.to_datetime("2017-09-19").date()
).sum()

53

53 late orders have the exact recorded customer delivery date of 19 september 2017

In [45]:
late_orders["order_delivered_customer_date"].dt.date.value_counts().head(10)

order_delivered_customer_date
2018-04-03    90
2018-04-10    82
2018-04-04    80
2018-03-23    76
2018-03-28    76
2018-04-11    74
2018-03-29    69
2018-04-02    67
2018-04-09    66
2018-04-16    65
Name: count, dtype: int64

In [46]:
(late_orders["days_late"]>60).sum()

79

so there are 79 of 6,534 late orders were more than 60 days late

In [47]:
extreme_late_rate = (
    (late_orders["days_late"]>60).sum()
    /late_orders.shape[0]
)*100
round(extreme_late_rate,2)

1.21

In [48]:
late_orders.sort_values(
    by = "order_delivered_customer_date",
    ascending = False
).head(1)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,days_late
18731,7e708aed151d6a8601ce8f2eaa712bf4,033fab69968b0d69099d64423831a236,delivered,2018-06-02 18:37:14,2018-06-02 18:51:31,2018-06-04 13:42:00,2018-10-17 13:22:46,2018-07-13,96


In [49]:
orders["order_delivered_customer_date"].max()

Timestamp('2018-10-17 13:22:46')

The order with the latest delivery date in the entire dataset 

In [50]:
orders["order_purchase_timestamp"].agg(["min","max"])

min   2016-09-04 21:15:19
max   2018-10-17 17:30:18
Name: order_purchase_timestamp, dtype: datetime64[ns]

Earliest purchase is 4 sep 2016
latest purchase 17 oct 2018

In [51]:
orders["order_id"].duplicated().sum()

0

In [52]:
orders["customer_id"].duplicated().sum()

0

In [53]:
orders.duplicated().sum()

0

Order Iems Table

In [54]:
order_items = pd.read_csv(
    "../01_raw_data/olist_order_items_dataset.csv"
)

In [55]:
order_items.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [56]:
order_items.shape

(112650, 7)

In [57]:
order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [58]:
order_items["shipping_limit_date"] = pd.to_datetime(
    order_items["shipping_limit_date"]
)
order_items.dtypes

order_id                       object
order_item_id                   int64
product_id                     object
seller_id                      object
shipping_limit_date    datetime64[ns]
price                         float64
freight_value                 float64
dtype: object

In [59]:
order_items.duplicated().sum()

0

In [60]:
order_items["order_id"].nunique()

98666

In [61]:
orders.loc[
    ~orders["order_id"].isin(order_items["order_id"]),
    "order_status"
].value_counts()

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

In [62]:
missing_item_orders = orders[
    ~orders["order_id"].isin(order_items["order_id"])
].copy()

In [63]:
missing_item_orders[
    missing_item_orders["order_status"] == "shipped"
]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
23254,a68ce1686d536ca72bd2dadc4b8671e5,d7bed5fac093a4136216072abaf599d5,shipped,2016-10-05 01:47:40,2016-10-07 03:11:22,2016-11-07 16:37:37,NaT,2016-12-01


In [64]:
order_items["order_id"].value_counts().head(10)

order_id
8272b63d03f5f79c56e9e4120aec44ef    21
1b15974a0141d54e36626dca3fdc731a    20
ab14fdcfbe524636d65ee38360e22ce8    20
9ef13efd6949e4573a18964dd1bbe7f5    15
428a2f660dc84138d969ccd69a0ab6d5    15
9bdc4d4c71aa1de4606060929dee888c    14
73c8ab38f07dc94389065f7eba4f297a    14
37ee401157a3a0b28c9c6d0ed8c3b24b    13
2c2a19b5703863c908512d135aa6accc    12
c05d6a79e55da72ca780ce90364abed9    12
Name: count, dtype: int64

The maximum order contains 21 item record

In [65]:
order_items[
    order_items["order_id"] == "8272b63d03f5f79c56e9e4120aec44ef"
]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
57297,8272b63d03f5f79c56e9e4120aec44ef,1,270516a3f41dc035aa87d220228f844c,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.2,7.89
57298,8272b63d03f5f79c56e9e4120aec44ef,2,05b515fdc76e888aada3c6d66c201dff,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.2,7.89
57299,8272b63d03f5f79c56e9e4120aec44ef,3,05b515fdc76e888aada3c6d66c201dff,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.2,7.89
57300,8272b63d03f5f79c56e9e4120aec44ef,4,05b515fdc76e888aada3c6d66c201dff,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.2,7.89
57301,8272b63d03f5f79c56e9e4120aec44ef,5,05b515fdc76e888aada3c6d66c201dff,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.2,7.89
57302,8272b63d03f5f79c56e9e4120aec44ef,6,05b515fdc76e888aada3c6d66c201dff,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.2,7.89
57303,8272b63d03f5f79c56e9e4120aec44ef,7,05b515fdc76e888aada3c6d66c201dff,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.2,7.89
57304,8272b63d03f5f79c56e9e4120aec44ef,8,05b515fdc76e888aada3c6d66c201dff,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.2,7.89
57305,8272b63d03f5f79c56e9e4120aec44ef,9,05b515fdc76e888aada3c6d66c201dff,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.2,7.89
57306,8272b63d03f5f79c56e9e4120aec44ef,10,05b515fdc76e888aada3c6d66c201dff,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.2,7.89


In [66]:
order_items[
    order_items["order_id"] == "8272b63d03f5f79c56e9e4120aec44ef"
]["product_id"].nunique()

3

In [67]:
order_items[
    order_items["order_id"] == "8272b63d03f5f79c56e9e4120aec44ef"
]["product_id"].value_counts()

product_id
270516a3f41dc035aa87d220228f844c    10
05b515fdc76e888aada3c6d66c201dff    10
79ce45dbc2ea29b22b5a261bbb7b7ee7     1
Name: count, dtype: int64

In [68]:
order_items[
    order_items["order_id"] == "8272b63d03f5f79c56e9e4120aec44ef"
]["seller_id"].nunique()

1

21 total units
3 unique products
1 seller

In [69]:
(
    order_items.groupby("order_id")["seller_id"].nunique() > 1
).sum()

1278

There are total 1278 multi seller orders. 

In [70]:
multi_seller_rate = (
    1278/order_items["order_id"].nunique()
)*100
round(multi_seller_rate,2)

1.3

In [71]:
order_items["seller_id"].nunique()

3095

there are total 3095 unique seller orders

In [72]:
order_items["product_id"].nunique()

32951

there are total 32951 unique productsin table

In [73]:
order_items["seller_id"].value_counts().head(10)

seller_id
6560211a19b47992c3666cc44a7e94c0    2033
4a3ca9315b744ce9f8e9374361493884    1987
1f50f920176fa81dab994f9023523100    1931
cc419e0650a3c5ba77189a1882b7556a    1775
da8622b14eb17ae2831f4ac5b9dab84a    1551
955fee9216a65b617aa5c0531780ce60    1499
1025f0e2d44d7041d6cf58b6550e0bfa    1428
7c67e1448b00f6e969d365cea6b010ab    1364
ea8482cd71df3c1969d7b9473ff13abc    1203
7a67c85e85bb2ce8582c35f2203ad736    1171
Name: count, dtype: int64

10 sellers with highest number of items sold

In [74]:
top_sellers = (
    order_items.groupby("seller_id")
    .agg(
        total_items=("product_id","count"),
        unique_products=("product_id","nunique")
    )
)
top_sellers["repeated_items"] = (
    top_sellers["total_items"] - 
    top_sellers["unique_products"]
)

top_sellers.sort_values(
    by = "total_items",
    ascending = False
).head(10)

,total_items,unique_products,repeated_items
seller_id,,,
6560211a19b47992c3666cc44a7e94c0,2033,256,1777
4a3ca9315b744ce9f8e9374361493884,1987,399,1588
1f50f920176fa81dab994f9023523100,1931,23,1908
cc419e0650a3c5ba77189a1882b7556a,1775,37,1738
da8622b14eb17ae2831f4ac5b9dab84a,1551,222,1329
955fee9216a65b617aa5c0531780ce60,1499,104,1395
1025f0e2d44d7041d6cf58b6550e0bfa,1428,154,1274
7c67e1448b00f6e969d365cea6b010ab,1364,198,1166
ea8482cd71df3c1969d7b9473ff13abc,1203,220,983


In [75]:
top_sellers["repeated_item_rate"] = (
    top_sellers["repeated_items"]/
    top_sellers["total_items"]
)*100
top_sellers.sort_values(
    by = "total_items",
    ascending = False
).head(10)

,total_items,unique_products,repeated_items,repeated_item_rate
seller_id,,,,
6560211a19b47992c3666cc44a7e94c0,2033,256,1777,87.407772
4a3ca9315b744ce9f8e9374361493884,1987,399,1588,79.919477
1f50f920176fa81dab994f9023523100,1931,23,1908,98.808907
cc419e0650a3c5ba77189a1882b7556a,1775,37,1738,97.915493
da8622b14eb17ae2831f4ac5b9dab84a,1551,222,1329,85.686654
955fee9216a65b617aa5c0531780ce60,1499,104,1395,93.062041
1025f0e2d44d7041d6cf58b6550e0bfa,1428,154,1274,89.215686
7c67e1448b00f6e969d365cea6b010ab,1364,198,1166,85.483871
ea8482cd71df3c1969d7b9473ff13abc,1203,220,983,81.712386


In [76]:
top_sellers["avg_units_per_product"] = (
    top_sellers["total_items"]/
    top_sellers["unique_products"]
)
top_sellers.sort_values(
    by = "total_items",
    ascending = False
).head(10)

,total_items,unique_products,repeated_items,repeated_item_rate,avg_units_per_product
seller_id,,,,,
6560211a19b47992c3666cc44a7e94c0,2033,256,1777,87.407772,7.941406
4a3ca9315b744ce9f8e9374361493884,1987,399,1588,79.919477,4.979950
1f50f920176fa81dab994f9023523100,1931,23,1908,98.808907,83.956522
cc419e0650a3c5ba77189a1882b7556a,1775,37,1738,97.915493,47.972973
da8622b14eb17ae2831f4ac5b9dab84a,1551,222,1329,85.686654,6.986486
955fee9216a65b617aa5c0531780ce60,1499,104,1395,93.062041,14.413462
1025f0e2d44d7041d6cf58b6550e0bfa,1428,154,1274,89.215686,9.272727
7c67e1448b00f6e969d365cea6b010ab,1364,198,1166,85.483871,6.888889
ea8482cd71df3c1969d7b9473ff13abc,1203,220,983,81.712386,5.468182


In [77]:
order_items["price"].describe()

count    112650.000000
mean        120.653739
std         183.633928
min           0.850000
25%          39.900000
50%          74.990000
75%         134.900000
max        6735.000000
Name: price, dtype: float64

In [78]:
order_items.sort_values(
    by = "price",
    ascending = False
).head(10)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
3556,0812eb902a67711a1cb742b3cdaa65ae,1,489ae2aa008f021502940f251d4cce7f,e3b4998c7a498169dc7bce44e6bb6277,2017-02-16 20:37:36,6735.00,194.31
112233,fefacc66af859508bf1a7934eab1e97f,1,69c590f7ffc7bf8db97190b6cb6ed62e,80ceebb4ee9b31afb6c6a916a574a1e2,2018-08-02 04:05:13,6729.00,193.21
107841,f5136e38d1a14a4dbd87dff67da82701,1,1bdf5e6731585cf01aa8169c7028d6ad,ee27a8f15b1dded4d213a468ba4eb391,2017-06-15 02:45:17,6499.00,227.66
74336,a96610ab360d42a2e5335a3998b4718a,1,a6492cc69376c469ab6f61d8f44de961,59417c56835dd8e2e72f91f809cd4092,2017-04-18 13:25:18,4799.00,151.34
11249,199af31afc78c699f0dbf71fb178d4d4,1,c3ed642d592594bb648ff4a04cee2747,59417c56835dd8e2e72f91f809cd4092,2017-05-09 15:50:15,4690.00,74.34
62086,8dbc85d1447242f3b127dda390d56e19,1,259037a6a41845e455183f89c5035f18,c72de06d72748d1a0dfb2125be43ba63,2018-06-28 12:36:36,4590.00,91.78
29193,426a9742b533fc6fed17d1fd6d143d7e,1,a1beef8f3992dbd4cd8726796aa69c53,512d298ac2a96d1931b6bd30aa21f61d,2018-08-16 14:24:28,4399.87,113.45
45843,68101694e5c5dc7330c91e1bbc36214f,1,6cdf8fc1d741c76586d8b6b15e9eef30,ed4acab38528488b65a9a9c603ff024a,2018-04-05 08:27:27,4099.99,75.27
78310,b239ca7cd485940b31882363b52e6674,1,dd113cb02b2af9c8e5787e8f1f0722f6,821fb029fc6e495ca4f08a35d51e53a5,2018-08-02 08:15:14,4059.00,104.51
59137,86c4eab1571921a6a6e248ed312f5a5a,1,6902c1962dd19d540807d0ab8fade5c6,fa1c13f2614d7b5c4749cbc52fecda94,2017-03-23 20:08:04,3999.90,17.01


In [79]:
q1 = order_items["price"].quantile(0.25)
q3 = order_items["price"].quantile(0.75)

iqr = q3 - q1
upper_price_limit = q3 + (1.5 * iqr)
upper_price_limit

277.4

Any item price above 277.40 would be flagged as a high price outlier

In [80]:
(order_items["price"] > upper_price_limit).sum()

8427

8427 item rows have a price greater than 277.40

In [81]:
order_items["freight_value"].describe()

count    112650.000000
mean         19.990320
std          15.806405
min           0.000000
25%          13.080000
50%          16.260000
75%          21.150000
max         409.680000
Name: freight_value, dtype: float64

In [82]:
(order_items["freight_value"] == 0).sum()

383

383 item rows have freight_value = 0

In [83]:
zero_freight_rate = (
    (order_items["freight_value"] == 0).sum()
    / order_items.shape[0]
)*100

round(zero_freight_rate, 2)

0.34

In [84]:
(order_items["freight_value"]>order_items["price"]).sum()

4124

4124 item rows have a freight cost higher than the product price

In [85]:
freight_higher_than_price_rate = (
    (order_items["freight_value"]>order_items["price"]).sum()
    /order_items.shape[0]
)*100
round(freight_higher_than_price_rate,2)

3.66

In [86]:
order_items["freight_price_ratio"] = (
    order_items["freight_value"]/
    order_items["price"]
)
order_items.sort_values(
    by = "freight_price_ratio",
    ascending = False
).head(10)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,freight_price_ratio
87081,c5bdd8ef3c0ec420232e668302179113,2,8a3254bee785a526d548a81a9bc3c9be,96804ea39d96eb908e7c3afdb671bb9e,2018-05-07 02:55:22,0.85,22.30,26.235294
27652,3ee6513ae7ea23bdfab5b9ab60bffcb5,1,8a3254bee785a526d548a81a9bc3c9be,96804ea39d96eb908e7c3afdb671bb9e,2018-05-04 03:55:26,0.85,18.23,21.447059
48625,6e864b3f0ec71031117ad4cf46b7f2a1,1,8a3254bee785a526d548a81a9bc3c9be,96804ea39d96eb908e7c3afdb671bb9e,2018-05-02 20:30:34,0.85,18.23,21.447059
110535,fb265b2dc558a56445dfc48f8224e201,1,baf25ed4f8f70238cc87230379471454,128f9bfbe4c7d5185033914b1de3d39a,2017-10-18 21:35:15,9.90,121.22,12.244444
94495,d642656598ae928a250620315d19e87e,1,b07fffe072c9adc235a35d8da7c0584d,dd533b429f380718b70ad9922c294bae,2018-03-13 10:55:37,4.99,37.04,7.422846
57304,8272b63d03f5f79c56e9e4120aec44ef,8,05b515fdc76e888aada3c6d66c201dff,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.20,7.89,6.575000
57309,8272b63d03f5f79c56e9e4120aec44ef,13,270516a3f41dc035aa87d220228f844c,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.20,7.89,6.575000
57308,8272b63d03f5f79c56e9e4120aec44ef,12,270516a3f41dc035aa87d220228f844c,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.20,7.89,6.575000
57307,8272b63d03f5f79c56e9e4120aec44ef,11,05b515fdc76e888aada3c6d66c201dff,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.20,7.89,6.575000
57306,8272b63d03f5f79c56e9e4120aec44ef,10,05b515fdc76e888aada3c6d66c201dff,2709af9587499e95e803a6498a5a56e9,2017-07-21 18:25:23,1.20,7.89,6.575000


In [87]:
products = pd.read_csv(
    "../01_raw_data/olist_products_dataset.csv"
)

products.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [88]:
products.shape

(32951, 9)

In [89]:
products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [90]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [91]:
products[
    products["product_category_name"].isnull()
][[
    "product_id",
    "product_category_name",
    "product_name_lenght",
    "product_description_lenght",
    "product_photos_qty"
]].head(10)

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty
105,a41e356c76fab66334f36de622ecbd3a,NaN,NaN,NaN,NaN
128,d8dee61c2034d6d075997acef1870e9b,NaN,NaN,NaN,NaN
145,56139431d72cd51f19eb9f7dae4d1617,NaN,NaN,NaN,NaN
154,46b48281eb6d663ced748f324108c733,NaN,NaN,NaN,NaN
197,5fb61f482620cb672f5e586bb132eae9,NaN,NaN,NaN,NaN
244,e10758160da97891c2fdcbc35f0f031d,NaN,NaN,NaN,NaN
294,39e3b9b12cd0bf8ee681bbc1c130feb5,NaN,NaN,NaN,NaN
299,794de06c32a626a5692ff50e4985d36f,NaN,NaN,NaN,NaN
347,7af3e2da474486a3519b0cba9dea8ad9,NaN,NaN,NaN,NaN
428,629beb8e7317703dcc5f35b5463fd20e,NaN,NaN,NaN,NaN


In [92]:
missing_product_info=products[
    products["product_category_name"].isnull()
]
missing_product_info[[
    "product_category_name",
    "product_name_lenght",
    "product_description_lenght",
    "product_photos_qty"
]].isnull().all(axis=1).sum()

610

610 products are missing for all four descriptive fields together

In [93]:
products[
    products["product_weight_g"].isnull()
]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
8578,09ff539a621711667c43eba6a3bd8466,bebes,60.0,865.0,3.0,NaN,NaN,NaN,NaN
18851,5eb564652db742ff8f28759cd8d2652a,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [94]:
order_items[
    order_items["product_id"].isin([
        "09ff539a621711667c43eba6a3bd8466",
        "5eb564652db742ff8f28759cd8d2652a"
    ])
]["product_id"].value_counts()
        

product_id
5eb564652db742ff8f28759cd8d2652a    17
09ff539a621711667c43eba6a3bd8466     1
Name: count, dtype: int64

This above output means those two products appear which have missing weight data, do appear in actual sales.

In [95]:
products["product_id"].duplicated().sum()

0

In [96]:
products.duplicated().sum()

0

In [97]:
products["product_category_name"].nunique()

73

There are 73 product categories with non missing category names

In [98]:
products["product_category_name"].value_counts().head(10)

product_category_name
cama_mesa_banho           3029
esporte_lazer             2867
moveis_decoracao          2657
beleza_saude              2444
utilidades_domesticas     2335
automotivo                1900
informatica_acessorios    1639
brinquedos                1411
relogios_presentes        1329
telefonia                 1134
Name: count, dtype: int64

In [99]:
products["product_weight_g"].describe()

count    32949.000000
mean      2276.472488
std       4282.038731
min          0.000000
25%        300.000000
50%        700.000000
75%       1900.000000
max      40425.000000
Name: product_weight_g, dtype: float64

In [100]:
(products["product_weight_g"] == 0).sum()

4

In [101]:
products[
    products["product_weight_g"] == 0
]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
9769,81781c0fed9fe1ad6e8c81fca1e1cb08,cama_mesa_banho,51.0,529.0,1.0,0.0,30.0,25.0,30.0
13683,8038040ee2a71048d4bdbbdc985b69ab,cama_mesa_banho,48.0,528.0,1.0,0.0,30.0,25.0,30.0
14997,36ba42dd187055e1fbe943b2d11430ca,cama_mesa_banho,53.0,528.0,1.0,0.0,30.0,25.0,30.0
32079,e673e90efa65a5409ff4196c038bb5af,cama_mesa_banho,53.0,528.0,1.0,0.0,30.0,25.0,30.0


The above is the 4 product whose weight is 0

In [102]:
zero_weight_products = products[
    products["product_weight_g"] == 0
]["product_id"]
order_items[
    order_items["product_id"].isin(zero_weight_products)
]["product_id"].value_counts()

product_id
e673e90efa65a5409ff4196c038bb5af    4
36ba42dd187055e1fbe943b2d11430ca    2
81781c0fed9fe1ad6e8c81fca1e1cb08    1
8038040ee2a71048d4bdbbdc985b69ab    1
Name: count, dtype: int64

4 products have weight 0. they appear in 8 sold item rows. these are likely data quality issues

In [103]:
(products[[
    "product_length_cm",
    "product_width_cm",
    "product_height_cm"
]] == 0).sum()

product_length_cm    0
product_width_cm     0
product_height_cm    0
dtype: int64

In [104]:
products["product_photos_qty"].describe()

count    32341.000000
mean         2.188986
std          1.736766
min          1.000000
25%          1.000000
50%          1.000000
75%          3.000000
max         20.000000
Name: product_photos_qty, dtype: float64

In [105]:
products["product_photos_qty"].value_counts().sort_index()

product_photos_qty
1.0     16489
2.0      6263
3.0      3860
4.0      2428
5.0      1484
6.0       968
7.0       343
8.0       192
9.0       105
10.0       95
11.0       46
12.0       35
13.0        9
14.0        5
15.0        8
17.0        7
18.0        2
19.0        1
20.0        1
Name: count, dtype: int64

In [106]:
missing_category_products = products[
    products["product_category_name"].isnull()
]["product_id"]

order_items[
    order_items["product_id"].isin(missing_category_products)
].shape[0]

1603

1603 item rows sold belong to products whose category is missing

In [107]:
missing_category_sales_rate=(
    1603/order_items.shape[0]
)*100
round(missing_category_sales_rate,2)

1.42

610 products have missing category information affecting 1603 item sales, 1.42% of all item rows

In [108]:
payments = pd.read_csv(
    "../01_raw_data/olist_order_payments_dataset.csv"
)
payments.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [109]:
payments.shape

(103886, 5)

In [110]:
payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [111]:
payments.duplicated().sum()

0

In [112]:
payments["order_id"].nunique()

99440

In [113]:
orders[
    ~orders["order_id"].isin(payments["order_id"])
]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
30710,bfbd0f9bdef84302105ad712db648a6c,86dc2ffce2dfff336de2f386a786e574,delivered,2016-09-15 12:16:38,2016-09-15 12:16:38,2016-11-07 17:11:53,2016-11-09 07:47:38,2016-10-04


In [114]:
order_items[
    order_items["order_id"] == "bfbd0f9bdef84302105ad712db648a6c"
]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,freight_price_ratio
84389,bfbd0f9bdef84302105ad712db648a6c,1,5a6b04657a4c5ee34285d1e4619a96b4,ecccfa2bb93b34a3bf033cc5d1dcdc69,2016-09-19 23:11:33,44.99,2.83,0.062903
84390,bfbd0f9bdef84302105ad712db648a6c,2,5a6b04657a4c5ee34285d1e4619a96b4,ecccfa2bb93b34a3bf033cc5d1dcdc69,2016-09-19 23:11:33,44.99,2.83,0.062903
84391,bfbd0f9bdef84302105ad712db648a6c,3,5a6b04657a4c5ee34285d1e4619a96b4,ecccfa2bb93b34a3bf033cc5d1dcdc69,2016-09-19 23:11:33,44.99,2.83,0.062903


one delivered containing three item records has no corresponding paymet record in the payment table

In [115]:
payments["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

In [116]:
payments[
    payments["payment_type"] == "not_defined"
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.0
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.0
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.0


In [117]:
orders[
    orders["order_id"].isin(
        payments.loc[
            payments["payment_type"] == "not_defined",
            "order_id"
        ]
    )
][[
    "order_id",
    "order_status",
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_customer_date"
]]

,order_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_customer_date
1130,00b1cb0320190ca0daa2c88b35206009,canceled,2018-08-28 15:26:39,NaT,NaT
39919,4637ca194b6387e2d538dc89b124b0ee,canceled,2018-09-03 14:14:25,NaT,NaT
40235,c8c528189310eaa44a745b8d9d26908b,canceled,2018-08-28 20:05:14,NaT,NaT


In [118]:
payments["payment_installments"].describe()

count    103886.000000
mean          2.853349
std           2.687051
min           0.000000
25%           1.000000
50%           1.000000
75%           4.000000
max          24.000000
Name: payment_installments, dtype: float64

In [119]:
payments[
    payments["payment_installments"] == 0
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
46982,744bade1fcf9ff3f31d860ace076d422,2,credit_card,0,58.69
79014,1a57108394169c0b47d8f876acc9ba2d,2,credit_card,0,129.94


In [120]:
payments[
    payments["order_id"].isin([
        "744bade1fcf9ff3f31d860ace076d422",
        "1a57108394169c0b47d8f876acc9ba2d"
    ])
].sort_values(
    by=["order_id","payment_sequential"]
)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
79014,1a57108394169c0b47d8f876acc9ba2d,2,credit_card,0,129.94
46982,744bade1fcf9ff3f31d860ace076d422,2,credit_card,0,58.69


In [121]:
first_payment_sequence = (
    payments.groupby("order_id")["payment_sequential"].min()
)
(first_payment_sequence > 1).sum () 

80

In [122]:
first_payment_sequence[
    first_payment_sequence > 1
].value_counts().sort_index()

payment_sequential
2    80
Name: count, dtype: int64

In [124]:
affected_orders = first_payment_sequence[
    first_payment_sequence > 1
].index
payments[
    payments["order_id"].isin(affected_orders)
].groupby("order_id").size().value_counts().sort_index()

1    78
2     2
Name: count, dtype: int64

In [126]:
payment_counts=(
    payments[
        payments["order_id"].isin(affected_orders)
    ]
    .groupby("order_id")
    .size()
)
two_payment_orders = payment_counts[
    payment_counts == 2
].index
payments[
    payments["order_id"].isin(two_payment_orders)
].sort_values(
    by = ["order_id", "payment_sequential"]
)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
36471,a079628ac8002126e75f86b0f87332e4,2,debit_card,1,50.00
10258,a079628ac8002126e75f86b0f87332e4,3,credit_card,3,102.82
60570,a4431cbd79dbddaae7988ce6091cbc3c,2,debit_card,1,26.20
90465,a4431cbd79dbddaae7988ce6091cbc3c,3,debit_card,1,20.00


80 orders have payment sequences beginning at 2 instead of 1. of these 78 have only sequence 2 available, while 2 orders have sequences 2 and 3.

In [127]:
payments["payment_value"].describe()

count    103886.000000
mean        154.100380
std         217.494064
min           0.000000
25%          56.790000
50%         100.000000
75%         171.837500
max       13664.080000
Name: payment_value, dtype: float64

In [128]:
(payments["payment_value"]==0).sum()

9

In [129]:
payments[
    payments["payment_value"] == 0
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
19922,8bcbe01d44d147f901cd3192671144db,4,voucher,1,0.0
36822,fa65dad1b0e818e3ccc5cb0e39231352,14,voucher,1,0.0
43744,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.0
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.0
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.0
62674,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.0
77885,fa65dad1b0e818e3ccc5cb0e39231352,13,voucher,1,0.0
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.0
100766,b23878b3e8eb4d25a158f57d96331b18,4,voucher,1,0.0


In [131]:
zero_voucher_orders = payments[
    (payments["payment_value"] == 0) &
    (payments["payment_type"] == "voucher")
]["order_id"].unique()

payments[
    payments["order_id"].isin(zero_voucher_orders)
].sort_values(
    by = ["order_id","payment_sequential"]
)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
33781,45ed6e85398a87c253db47c2d9f48216,1,voucher,1,21.13
11755,45ed6e85398a87c253db47c2d9f48216,2,voucher,1,50.01
62674,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.00
40546,6ccb433e00daae1283ccc956189c82ae,1,credit_card,5,84.67
93478,6ccb433e00daae1283ccc956189c82ae,2,voucher,1,14.65
92318,6ccb433e00daae1283ccc956189c82ae,3,voucher,1,22.72
43744,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.00
20963,8bcbe01d44d147f901cd3192671144db,1,credit_card,1,36.21
63762,8bcbe01d44d147f901cd3192671144db,2,voucher,1,31.70
5163,8bcbe01d44d147f901cd3192671144db,3,voucher,1,6.25


In [132]:
payments["order_id"].value_counts().head(10)

order_id
fa65dad1b0e818e3ccc5cb0e39231352    29
ccf804e764ed5650cd8759557269dc13    26
285c2e15bebd4ac83635ccc563dc71f4    22
895ab968e7bb0d5659d16cd74cd1650c    21
fedcd9f7ccdc8cba3a18defedd1a5547    19
ee9ca989fc93ba09a6eddc250ce01742    19
4bfcba9e084f46c8e3cb49b0fa6e6159    15
21577126c19bf11a0b91592e5844ba78    15
3c58bffb70dcf45f12bdf66a3c215905    14
4689b1816de42507a7d63a4617383c59    14
Name: count, dtype: int64

In [133]:
payments.groupby("payment_type")["payment_installments"].max()

payment_type
boleto          1
credit_card    24
debit_card      1
not_defined     1
voucher         1
Name: payment_installments, dtype: int64

In [134]:
payments[
    payments["payment_installments"] == 24
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
2970,70b7e94ea46d3e8b5bc12a50186edaf0,1,credit_card,24,274.84
10791,859f516f2fc3f95772e63c5757ab0d5b,1,credit_card,24,609.56
12307,ff36cbc44b8f228e0449c92ef089c843,1,credit_card,24,756.49
18512,2b7dbe9be72b8f9733844c31055c0825,1,credit_card,24,345.39
21713,6ae2e8b8fac02522481d2a2f4ca4412c,1,credit_card,24,433.43
23024,90f864fe19d11549fa01eb81c4dd87e3,1,credit_card,24,588.58
36088,84d2098c97827c6327ed4d7be95e1fc8,1,credit_card,24,286.78
50401,ffb18bf111fa70edf316eb0390427986,1,credit_card,24,617.24
52846,63dbe0c8e63e5f1b4deec09d4f044a7f,1,credit_card,24,771.69
55094,fcbb6af360b31b05460c2c8e524588c0,1,credit_card,24,1194.38


In [135]:
payments[
    payments["payment_type"] == "credit_card"
]["payment_installments"].value_counts().sort_index()

payment_installments
0         2
1     25455
2     12413
3     10461
4      7098
5      5239
6      3920
7      1626
8      4268
9       644
10     5328
11       23
12      133
13       16
14       15
15       74
16        5
17        8
18       27
20       17
21        3
22        1
23        1
24       18
Name: count, dtype: int64

In [136]:
payments.sort_values(
    by="payment_value",
    ascending=False
).head(10)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
52107,03caa2c082116e1d31e67e9ae3700499,1,credit_card,1,13664.08
34370,736e1922ae60d0d6a89247b851902527,1,boleto,1,7274.88
41419,0812eb902a67711a1cb742b3cdaa65ae,1,credit_card,8,6929.31
49581,fefacc66af859508bf1a7934eab1e97f,1,boleto,1,6922.21
85539,f5136e38d1a14a4dbd87dff67da82701,1,boleto,1,6726.66
62409,2cc9089445046817a7539d90805e6e5a,1,boleto,1,6081.54
43232,a96610ab360d42a2e5335a3998b4718a,1,credit_card,10,4950.34
70320,b4c4b76c642808cbe472a32b86cddc95,1,credit_card,5,4809.44
6440,199af31afc78c699f0dbf71fb178d4d4,1,credit_card,8,4764.34
67546,8dbc85d1447242f3b127dda390d56e19,1,credit_card,8,4681.78


In [137]:
order_items[
    order_items["order_id"] == "03caa2c082116e1d31e67e9ae3700499"
]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,freight_price_ratio
1647,03caa2c082116e1d31e67e9ae3700499,1,5769ef0a239114ac3a854af00df129e4,b37c4c02bda3161a7546a4e6d222d5b2,2017-10-06 15:28:20,1680.0,28.01,0.016673
1648,03caa2c082116e1d31e67e9ae3700499,2,5769ef0a239114ac3a854af00df129e4,b37c4c02bda3161a7546a4e6d222d5b2,2017-10-06 15:28:20,1680.0,28.01,0.016673
1649,03caa2c082116e1d31e67e9ae3700499,3,5769ef0a239114ac3a854af00df129e4,b37c4c02bda3161a7546a4e6d222d5b2,2017-10-06 15:28:20,1680.0,28.01,0.016673
1650,03caa2c082116e1d31e67e9ae3700499,4,5769ef0a239114ac3a854af00df129e4,b37c4c02bda3161a7546a4e6d222d5b2,2017-10-06 15:28:20,1680.0,28.01,0.016673
1651,03caa2c082116e1d31e67e9ae3700499,5,5769ef0a239114ac3a854af00df129e4,b37c4c02bda3161a7546a4e6d222d5b2,2017-10-06 15:28:20,1680.0,28.01,0.016673
1652,03caa2c082116e1d31e67e9ae3700499,6,5769ef0a239114ac3a854af00df129e4,b37c4c02bda3161a7546a4e6d222d5b2,2017-10-06 15:28:20,1680.0,28.01,0.016673
1653,03caa2c082116e1d31e67e9ae3700499,7,5769ef0a239114ac3a854af00df129e4,b37c4c02bda3161a7546a4e6d222d5b2,2017-10-06 15:28:20,1680.0,28.01,0.016673
1654,03caa2c082116e1d31e67e9ae3700499,8,5769ef0a239114ac3a854af00df129e4,b37c4c02bda3161a7546a4e6d222d5b2,2017-10-06 15:28:20,1680.0,28.01,0.016673


In [138]:
reviews = pd.read_csv(
    "../01_raw_data/olist_order_reviews_dataset.csv"
)

reviews.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [147]:
reviews.shape

(99224, 7)

In [148]:
reviews["review_id"].nunique()

98410

In [149]:
review_id_counts = reviews["review_id"].value_counts()
review_id_counts[
    review_id_counts > 1
].value_counts().sort_index()

count
2    764
3     25
Name: count, dtype: int64

In [150]:
reviews["order_id"].nunique()

98673

There are total 98673 unique order_id

In [151]:
order_review_counts = reviews["order_id"].value_counts()
order_review_counts[
    order_review_counts > 1
].value_counts().sort_index()

count
2    543
3      4
Name: count, dtype: int64

In [153]:
two_review_orders = order_review_counts[
    order_review_counts == 2
].index
two_review_rows = reviews[
    reviews["order_id"].isin(two_review_orders)
]
two_review_rows.shape

(1086, 7)

In [154]:
two_review_summary = (
    two_review_rows
    .groupby("order_id")
    .agg(
        review_rows=("review_id","size"),
        unique_review_ids=("review_id","nunique")
    )
)
two_review_summary["unique_review_ids"].value_counts()

unique_review_ids
2    543
Name: count, dtype: int64

so 543 orders has 2 different id's not the duplicate one

In [156]:
three_review_orders = order_review_counts[
    order_review_counts == 3
].index
three_review_rows = reviews[
    reviews["order_id"].isin(three_review_orders)
]
three_review_rows.shape

three_review_summary = (
    three_review_rows
    .groupby("order_id")
    .agg(
        review_rows=("review_id","size"),
        unique_review_ids=("review_id","nunique")
    )
)
three_review_summary["unique_review_ids"].value_counts()

unique_review_ids
3    4
Name: count, dtype: int64

In [158]:
three_review_rows[[
    "order_id",
    "review_id",
    "review_score",
    "review_comment_title",
    "review_comment_message",
    "review_creation_date",
    "review_answer_timestamp"
]].sort_values(
    by = ["order_id","review_creation_date"]
)

,order_id,review_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
69438,03c939fd7fd3b38f8485a0f95798f1f6,405eb2ea45e1dbe2662541ae5b47e2aa,3,NaN,Seria ótimo se tivesem entregue os 3 (três) pe...,2018-03-06 00:00:00,2018-03-06 19:50:32
8273,03c939fd7fd3b38f8485a0f95798f1f6,b04ed893318da5b863e878cd3d0511df,3,NaN,Um ponto negativo que achei foi a cobrança de ...,2018-03-20 00:00:00,2018-03-21 02:28:23
51527,03c939fd7fd3b38f8485a0f95798f1f6,f4bb9d6dd4fb6dcc2298f0e7b17b8e1e,4,NaN,NaN,2018-03-29 00:00:00,2018-03-30 00:29:09
44694,8e17072ec97ce29f0e1f111e598b0c85,67c2557eb0bd72e3ece1e03477c9dff5,1,NaN,Entregou o produto errado.,2018-04-07 00:00:00,2018-04-08 22:48:27
64510,8e17072ec97ce29f0e1f111e598b0c85,2d6ac45f859465b5c185274a1c929637,1,NaN,Comprei 3 unidades do produto vieram 2 unidade...,2018-04-07 00:00:00,2018-04-07 21:13:05
92300,8e17072ec97ce29f0e1f111e598b0c85,6e4c4086d9611ae4cc0cc65a262751fe,1,NaN,"Embora tenha entregue dentro do prazo, não env...",2018-04-14 00:00:00,2018-04-16 11:37:31
89360,c88b1d1b157a9999ce368f218a407141,fb96ea2ef8cce1c888f4d45c8e22b793,5,NaN,NaN,2017-07-21 00:00:00,2017-07-26 13:45:15
1985,c88b1d1b157a9999ce368f218a407141,ffb8cff872a625632ac983eb1f88843c,3,NaN,NaN,2017-07-22 00:00:00,2017-07-26 13:41:07
82525,c88b1d1b157a9999ce368f218a407141,202b5f44d09cd3cfc0d6bd12f01b044c,5,NaN,NaN,2017-07-22 00:00:00,2017-07-26 13:40:22
13982,df56136b8031ecd28e200bb18e6ddb2e,72a1098d5b410ae50fbc0509d26daeb9,5,NaN,NaN,2017-02-07 00:00:00,2017-02-10 10:46:09


An order can have multiple distinct reviews in this dataset. Repeated order_id values in Reviews should not automatically be deleted as duplicates 

In [159]:
reviews["review_score"].value_counts().sort_index()

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64

In [160]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [161]:
reviews.duplicated().sum()

0

In [162]:
reviews["review_creation_date"] = pd.to_datetime(
    reviews["review_creation_date"]
)
reviews["review_answer_timestamp"] = pd.to_datetime(
    reviews["review_answer_timestamp"]
)
reviews.dtypes

review_id                          object
order_id                           object
review_score                        int64
review_comment_title               object
review_comment_message             object
review_creation_date       datetime64[ns]
review_answer_timestamp    datetime64[ns]
dtype: object

In [163]:
sellers = pd.read_csv(
    "../01_raw_data/olist_sellers_dataset.csv"
)

sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [164]:
sellers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [168]:
sellers["seller_zip_code_prefix"] = (
    sellers["seller_zip_code_prefix"]
    .astype("string")
    .str.zfill(5)
)

In [169]:
sellers["seller_id"].duplicated().sum()

0

In [170]:
geolocation = pd.read_csv(
    "../01_raw_data/olist_geolocation_dataset.csv"
)

geolocation.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [171]:
geolocation["geolocation_zip_code_prefix"] = (
    geolocation["geolocation_zip_code_prefix"]
    .astype("string")
    .str.zfill(5)
)

In [172]:
geolocation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  string 
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), object(2), string(1)
memory usage: 38.2+ MB


In [173]:
geolocation["geolocation_zip_code_prefix"].nunique()

19015

so 19015 unique postcode prefix

In [174]:
geolocation.duplicated().sum()

261831

In [175]:
(
    geolocation.groupby("geolocation_zip_code_prefix")["geolocation_state"]
    .nunique()
    .gt(1)
    .sum()
)

8

8 out of 19,015 unique postcode prefixes are linked to more than one state.

In [181]:
(
    ~sellers["seller_zip_code_prefix"]
    .isin(["geolocation_zip_code_prefix"])
).sum()

3095

In [178]:
print("Seller postcode dtype:",
      sellers["seller_zip_code_prefix"].dtype)

print("Geolocation postcode dtype:",
      geolocation["geolocation_zip_code_prefix"].dtype)

Seller postcode dtype: string
Geolocation postcode dtype: string


In [180]:
first_seller_zip = sellers.loc[0, "seller_zip_code_prefix"]
print("seller postcode:", repr(first_seller_zip))
print(
    "Matches in geolocation",
    (geolocation["geolocation_zip_code_prefix"] == first_seller_zip).sum()
)

seller postcode: '13023'
Matches in geolocation 80


Seller and geolocation postcode formats can match successfully 

In [182]:
(
    ~sellers["seller_zip_code_prefix"]
    .isin(geolocation["geolocation_zip_code_prefix"])
).sum()

7

In [184]:
(
    ~customers["customer_zip_code_prefix"]
    .astype("string")
    .str.zfill(5)
    .isin(geolocation["geolocation_zip_code_prefix"])
).sum()

278

278 customer records that have postcode prefixes that do not exist in the geolocation table out of 99,163 customer records

In [185]:
category_translation = pd.read_csv(
    "../01_raw_data/product_category_name_translation.csv"
)

category_translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [186]:
category_translation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB


In [188]:
missing_translation = set(
    products["product_category_name"].dropna().unique()
) - set(
    category_translation["product_category_name"].unique()
)
missing_translation

{'pc_gamer', 'portateis_cozinha_e_preparadores_de_alimentos'}